### ISSUE TITLE
intro from 'where time stood still' with bottom line(s) cut off

### ISSUE THREAD

**Issue by mithrendal** [dirkwhoffmann/vAmiga#743]
this issue has EAB member [Photon](http://eab.abime.net/member.php?u=3981) reported to me

Photon wrote this:
`The emulation seems to limit the Copper display to ~200 lines or else start emulation in NTSC by default. (NTSC option is off. I was warned by a bootblock detecting such.)`

He wrote on his A500 he can see the missing lines 

![image](https://user-images.githubusercontent.com/17108995/196390821-2797fd02-7fed-494b-bf58-0754a7704583.png)

I tested it in an older vAmigaWeb Release V1.08 where it also exhibits the missing lines...

**Comment by dirkwhoffmann** [comment]
Strange enough. The frame type is reported as PAL although I am pretty sure it's NTSC related. Needs investigation...

<img width="956" alt="Bildschirmfoto 2022-10-18 um 11 37 52" src="https://user-images.githubusercontent.com/12561945/196395462-32ba56b4-8f32-4c62-bbde-59cf1356302f.png">
X-ray (Sprites and Blitter hidden): 
<img width="883" alt="Bildschirmfoto 2022-10-18 um 06 39 44" src="https://user-images.githubusercontent.com/12561945/196395643-467bc0ae-6739-42f1-a061-2a099393f01f.png">

**Comment by dirkwhoffmann** [comment]
Running with `BORDER_DEBUG = 1` reveals that the DIW is too small. 

<img width="956" alt="Bildschirmfoto 2022-10-18 um 11 59 56" src="https://user-images.githubusercontent.com/12561945/196400673-515ae964-6a04-4c9f-b948-5c0cca1cd591.png">

**Comment by mras0** [comment]
From https://eab.abime.net/showpost.php?p=414480&postcount=18:
> Write to BPL0DAT enables bitplane shift registers inside Denise (this is nothing new, Agnus DMA does this automatically internally). My guess is that there is logic that outputs background color as long as BPL0DAT has not been written to.
> Test used: DDFSTRT set to 0x80, put sprite on left side of screen. It is not visible. Use copper to write to BPL0DAT on left side of screen. Single raster line of sprite is now visible + 16 pixels of bitplane data that was written to BPL0DAT.

In the following post Photon speculates how it can be used :)

**Comment by mithrendal** [comment]
Ah ok that’s probably what Photon means with `copper display`  and limit `the Copper display to ~200 lines` …

**Comment by dirkwhoffmann** [comment]
> Ah ok that’s probably what Photon means with copper display and limit the Copper display to ~200 lines

I don't get it. To me it seems like the DIW is set incorrectly (maybe as the result of a failed PAL detection).

**Comment by mithrendal** [comment]
After reading the article which @mras0 points to and the comments from photon who apparently wrote the intro. I guessed (by mixing these two stories) that it might be some copper instructions with these bpl0dst writes which after line 200 do not succeed in the desired result. My guess in understanding of the two statements could be completely wrong.

**Comment by mras0** [comment]
Might be the wrong undocumented feature I linked to, but it seems like the display can be activated outside the normal display window. Quick test case: [foo.zip](https://github.com/dirkwhoffmann/vAmiga/files/9819463/foo.zip)

In Winuae (A500 quickstart):
![image](https://user-images.githubusercontent.com/9952017/196663882-af67a931-747f-482a-83c4-7766dc9ae539.png)

Don't have any OCS/ECS machine to test, but it works on my a1200 (though display looks different).

**Comment by dirkwhoffmann** [comment]
Nice test! I'll add it to the test repo...

**Comment by mithrendal** [comment]
@mras0 you are best 😀

![C5590481-1707-41D6-AEEE-15965928AF07](https://user-images.githubusercontent.com/17108995/196745919-ff474cfa-ae61-427c-a04b-830d9a59f1d5.png)

running foo in current vAmiga

**Comment by HenrikErlandsson** [comment]
Oh, I see now. Only a few demos use sprites in the "border". Simplified: Normally the sprite engine stops bitplane DMA at DIWSTOP, and sprite bitplane DMA output is stopped with it. The Copper commands extend the bitplane DMA so the stars can fill the screen and go overscan.

This required an update to WinUAE some years ago. But I think it's an edge case and it's OK to go by DIWSTOP (as long as e.g. color changes take effect. Quite a few releases put background color changes outside the playfield.)

But I wrote [Sprite Programming](http://coppershade.org/articles/AMIGA/Denise/Sprite_Programming/) a while back to cover all cases regarding the sprite engine's behavior on OCS and up (but not covering ECS/AGA added features much). Maybe there is still something useful there to check against, without having to cater to "border" tricks.

**Comment by dirkwhoffmann** [comment]
OK, I think I got it now. My confusion was mostly because I didn't realize that everything except the spaceship is drawn with sprites. Disabling sprites in RetroShell with 

    denise hide sprites 0xFF

reveals it (DMA debugger shows Copper and bitplane DMA):

<img width="964" alt="Bildschirmfoto 2022-10-25 um 14 12 18" src="https://user-images.githubusercontent.com/12561945/197770652-e4c8d8fa-a320-4b10-a3e3-63d359c919dd.png">

**Comment by mras0** [comment]
Yes, perhaps I muddied it up with my testcase by using BPL1DAT writes to also fill screen (wanted to see how that worked, the intro just writes 0 to enable sprites).

Denise probably doesn't know about the y-part of DIWSTRT/DIWSTOP and just enables output on BPL1DAT write if it happens late enough, doing one write with copper while waiting for HPOS $27 doesn't enable sprites while $29 does. Don't know if that's a fixed position (doesn't seem to be related to x-part of DIWSTRT).

From a brief look at the source it looks like Agnus::executeUntil is currently driving the drawing? That probably has to be changed a bit.

**Comment by dirkwhoffmann** [comment]
Brief update: The major issue seems to be how and when the border is drawn. I.e., when the border is drawn prior to the sprites by changing this

```
        // Draw sprites
        drawSprites();

        // Perform playfield-playfield collision check (if enabled)
        if (config.clxPlfPlf) checkP2PCollisions();

        // Draw border pixels
        drawBorder();
```

into that

```
        // Draw border pixels
        drawBorder();

        // Draw sprites
        drawSprites();

        // Perform playfield-playfield collision check (if enabled)
        if (config.clxPlfPlf) checkP2PCollisions();
```
 
then everything looks fine:
 
<img width="878" alt="Bildschirm­foto 2022-10-29 um 15 11 56" src="https://user-images.githubusercontent.com/12561945/198833695-e7d18e9e-eef0-4470-ab4c-5ee4444cf50a.png">

**Comment by dirkwhoffmann** [comment]
Final findings: vAmiga did everything right internally. The bug was due to some extra logic  ensuring that nothing is drawn inside the vertical border area (variable `lineWasBlank`). The fix is simple: In `Denise::drawBorder`, line

    if (agnus.sequencer.lineWasBlank || (!flop && !on)) {

has to be replaced by 

    if (!flop && !on) {

This also fixes the test written by @mras0: 
<img width="851" alt="Bildschirm­foto 2022-11-05 um 09 35 58" src="https://user-images.githubusercontent.com/12561945/200111357-ec65c168-2c2e-4488-b931-e568a036b5cc.png">
The test is part of the vAmigaTS repo now. It is located at `Agnus/DIW/DIWV/diwvmras`.

**Comment by dirkwhoffmann** [comment]
Fixed in v2.2b1
